tests/unit/_modules/test_packagemanager.py

```diff
@@ -0,0 +1,345 @@
+    '''
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    def test_PackageManager_opensuse(self, dist_return):
+        """
+        Test Packagemanager assignments based on `patform` mocks.
+        """
+        dist_return.return_value = ('opensuse', 42.1, 'x86_64')
+        ret = PackageManager()
+        assert isinstance(ret.pm, Zypper) is True
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    def test_PackageManager_suse(self, dist_return):
+        """
+        Test Packagemanager assignments based on `patform` mocks.
+        """
+        dist_return.return_value = ('SUSE', '12.2', 'x86_64')
+        ret = PackageManager()
+        assert isinstance(ret.pm, Zypper) is True
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    def test_PackageManager_centos(self, dist_return):
+        """
+        Test Packagemanager assignments based on `patform` mocks.
+        """
+        dist_return.return_value = ('centos', '8', 'x86_64')
+        ret = PackageManager()
+        assert isinstance(ret.pm, Apt) is True
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    def test_PackageManager_fedora(self, dist_return):
+        """
+        Test Packagemanager assignments based on `patform` mocks.
+        """
+        dist_return.return_value = ('fedora', '23', 'x86_64')
+        ret = PackageManager()
+        assert isinstance(ret.pm, Apt) is True
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    @mock.patch('srv.salt._modules.packagemanager.Popen')
+    def test_reboot(self, po, dist_return):
+        """
+        Reboot method
+        """
+        dist_return.return_value = ('opensuse', '42.2', 'x86_64')
+        ret = PackageManager()
+        ret.reboot_in()
+        assert po.called is True
+
+    @mock.patch('srv.salt._modules.packagemanager.linux_distribution')
+    @mock.patch('srv.salt._modules.packagemanager.Popen')
+    def test_not_implemented(self, po, dist_return):
+        """
+        Your platform is not supported
+        """
+        dist_return.return_value = ('ScientificLinux', '42.2', 'x86_64')
+        with pytest.raises(ValueError) as excinfo:
+            PackageManager()
+        excinfo.match('Failed to detect PackageManager for OS.*')
+
+class TestZypper():
+
+    @pytest.fixture(scope='class')
+    def zypp(self):
+        """
+        Fixture to always get Zypper.
+        """
+        self.linux_dist = patch('srv.salt._modules.packagemanager.linux_distribution')
+        self.lnx_dist_object  = self.linux_dist.start()
+        self.lnx_dist_object.return_value = ('opensuse', '42.2', 'x86_64')
+        args = {'debug': False, 'kernel': False, 'reboot': False}
+        yield PackageManager(**args).pm
+        self.linux_dist.stop()
+
+    @mock.patch('srv.salt._modules.packagemanager.Popen')
+    def test__refresh(self, po, zypp):
```

Comment:
I know that it's not __needed__ to test private methods, as you only want to validate the "output". I think in our case though it's not enough to validate the "output" because `_handle` can take multiple routes and calls methods on it's way that will try to bring the system in a certain state.
In general, I agree with the "only test the public API" pragma, in this case though I think that we should also validate the underlying private methods.

feedback/counter-arguments welcome